CSharp/Library/Microsoft.Bot.Connector.Shared/MicrosoftAppCredentials.cs

```diff
@@ -266,6 +284,14 @@ private bool TokenNotExpired(OAuthResponse token)
             return token.expiration_time > DateTime.UtcNow;
         }
 
+        private bool TokenHalfwayExpired(OAuthResponse token,  int secondsToExpire = 60)
+        {
+            int secondsToHalfwayExpire = Math.Min(token.expires_in / 2, 1800);
+            TimeSpan TimeToExpiration = token.expiration_time - DateTime.UtcNow;
+            return TimeToExpiration.TotalSeconds < secondsToHalfwayExpire
+                && TimeToExpiration.TotalSeconds > secondsToExpire;

```

Comment:
if TotalSeconds calculated is < 60 then this will be false.

Reply:
Is not that the intended behavior? If TotalSeconds < 60 then Token is not halfway expired, it's expired. It should return false for this one and will be true for TokenExpired.